posts/frangeris/04.upgradeable-contracts.md

```diff
@@ -0,0 +1,228 @@
+Welcome to Node.js v14.18.1.
+Type ".help" for more information.
+> const proxy = await ethers.getContractFactory("Gateway")
+undefined
+> const gateway = await proxy.attach("0xe7197e1D5F60eE3745a828a4CeaDAA0FBfFF886D")
+undefined
+> (await gateway.getFee()).toString()
+'1'
+>
+```
+
+As you can see, we are able to interact with the `Gateway` contract via the `TransparentUpgradeableProxy` contract, but what if we want to upgrade the `Gateway` contract? Let's add support to allows the owner to change the fee, so let's create `GatewayV2.sol`:
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+contract GatewayV2 is Initializable, OwnableUpgradeable {
+    uint private fee;
+
+    function initialize(uint _fee) public initializer {
+        fee = _fee;
+        __Ownable_init();
+    }
+
+    function getFee() public view returns (uint) {
+        return fee;
+    }
+
+    function setFee(uint _fee) public onlyOwner {
+        fee = _fee;
+    }
+}
+```
+
+The new contract should be exactly the same as the previous one, but with the addition of the `setFee` function, now let's deploy it:
+
+> The difference between `upgrade` script and the previous one is that we are using the `upgradeProxy` function instead of `deployProxy`, this function will upgrade the `Gateway` contract to the new `GatewayV2` contract.
+
+```bash
+frang @ ~/Code/posts-examples/payments-gateway (master)
+ →  PROXY_ADDRESS=0xe7197e1D5F60eE3745a828a4CeaDAA0FBfFF886D npx hardhat run scripts/upgrade.js --network sepolia
+Upgrading...
+Upgraded: 0xe7197e1D5F60eE3745a828a4CeaDAA0FBfFF886D
+```
+
+The contract was upgraded, but how can we verify that? Let's try to interact with the contract again BUT this time we will use the `GatewayV2` contract instead of the `Gateway` contract while using the same contract address:
+
+```bash
+frang @ ~/Code/posts-examples/payments-gateway (master)
+ →  npx hardhat console --network sepolia
+Welcome to Node.js v14.18.1.
+Type ".help" for more information.
+> const proxy = await ethers.getContractFactory("GatewayV2")
+undefined
+> const gateway = await proxy.attach("0xe7197e1D5F60eE3745a828a4CeaDAA0FBfFF886D")
+undefined
+> (await gateway.getFee()).toString()
+'1'
+> await gateway.setFee([2])
+{
+  hash: '0x699ad32e17ca56537d69c60ff4727526f031f53ca46775dad535d37300055cda',
+  ...
+}
+> (await gateway.getFee()).toString()
+'2'
+```
+
+The contract now has the `setFee` function, so we can change the fee, but how can we verify that the contract was upgraded? Let's take a look at the transaction details on [Sepolia Block Explorer](https://sepolia.etherscan.io/tx/0xc0ea9e236e0aea3534bd060fd081468c39649d0fb00cb44b19e241c85338757b#eventlog), you should be able to see the `Upgraded` event, this event is emitted when the `Gateway` contract was upgraded to the `GatewayV2` contract.
```

Comment:
I believe we've already told how to check if the contract was actually upgraded. Are we checking a different thing this time?

Reply:
No, an event in the blockchain is the main source of truth to really assure that the upgrade method was executed, also is important to understand and look for events on blockchain logs

Reply:
The first way to check if the contract was upgraded is by calling the new method, but after that, we also have another way to check that it was upgraded, this way consists in checking inside the blockchain, that's why